doc/compiler/il/SymbolsSymrefsAliasing.md

```diff
@@ -0,0 +1,172 @@
+# Symbols, Symrefs, and Aliasing in Testarossa
+
+## Introduction
+
+This document is an initial attempt to explain and document the concepts and
+implementation of symbols, symbols references, and aliasing in Testarossa.
+
+## Symbols and symrefs
+
+The original, conceptual distinction between symbols and symbol references (symrefs)
+was actually quite straightforward. However, over time, implementation details have
+lead to the definition of the two to become very blurry and somewhat unspecified.
+This document therefore offers the original definitions, as well as some implementation
+details and subtleties.
+
+### Symbols
+
+A symbol represents a memory location. It holds the minimum information required to
+directly translate a method's source code or IL representation, without consideration
+for optimization. It stores basic information (e.g. data type) needed to determine
+what memory access operations can/must be done.
+
+Different kinds of symbols exist to represent the different kinds of memory access
+requirements. Each kind of symbol stores information that is specific to its kind (type).
+The different symbol kinds are:
+
+* statics
+   * represent static memory address
+   * store the address of the memory location
+* autos
+   * represent variables mapped on the stack
+      * may come from local method variables or JIT allocated stack objects
+   * store stack offset of the variable
+* methods
+   * represent a method (static, virtual)
+* resolved method
+* parameter
+* register mapped
+* label
+* shadow
+   * represent indirection off a base address
+      * e.g. accessing a class field using the classes base address plus an offset
+   * store offset from some base address
+
+#### int shadow
+
+An int shadow represents the common case of an integer type. In the case of Java,
+for example, an int shadow is created for a field or array access.
+
+#### generic int shadow
+
+A generic int shadow is created by the optimizer in (relatively rare) cases when we
+want to refer to a field that wasn't referred to in the bytecodes. If the optimizer
+wants to synthesize a load or store, but doesn't have a symref for it in the original
+program, it can create one and call it a generic int shadow.
+
+For example, escape analysis may want to generate a store to a location within an
+object that hasn't been read in this method in order to initialize the location to zero.
```

Comment:
Come to think of it, in light of that perhaps its better worded as "For example, an escape analysis optimization might..."